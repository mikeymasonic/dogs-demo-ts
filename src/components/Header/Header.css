.Header {
  display: flex;
  justify-content: space-between;
}

.Header h1 {
  text-decoration: none;
  margin-top: -0.15rem;
  cursor: pointer;
  color: gray;
  -webkit-text-stroke: 1px black;
}

.Header h1:hover {
  color: hotpink;
}

.Header input {
  width: 15rem;
  height: 2rem;
  margin-top: -0.25rem;
}
