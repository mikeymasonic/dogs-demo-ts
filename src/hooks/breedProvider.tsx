import React, { createContext, useState, useContext, useEffect } from 'react';
import { fetchList, fetchDetail } from '../services/dogsAPI';

export type BreedContext = {
  breeds: Array<string>;
  filteredBreeds: Array<string>;
  searchText: string;
  breedDetail: Array<string>;
  loading: boolean;
  handleChange: (a: { target: any }) => void;
  handleDetail: (breed: string) => void;
};

const BreedContext = createContext<BreedContext | null>(null);

export const BreedProvider = ({ children }: { children: React.ReactNode }) => {
  const [breeds, setBreeds] = useState<string[]>([]);
  const [filteredBreeds, setFilteredBreeds] = useState<string[]>([]);
  const [searchText, setSearchText] = useState('');
  const [breedDetail, setBreedDetail] = useState<string[]>([]);
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    setLoading(true);
    fetchList()
      .then((apiBreeds) => setBreeds(apiBreeds))
      .finally(() => setLoading(false));
  }, []);

  useEffect(() => {
    setFilteredBreeds(breeds.slice(0, 12));
  }, [breeds]);

  useEffect(() => {
    const searchResults = breeds.filter((breed) => breed.includes(searchText));
    setLoading(true);
    setFilteredBreeds(searchResults.slice(0, 12));
    setLoading(false);
  }, [breeds, searchText]);

  const handleChange = ({ target }: { target: any }) => {
    setSearchText(target.value);
  };

  const handleDetail = (breed: string) => {
    setLoading(true);
    fetchDetail(breed)
      .then((apiBreeds) => setBreedDetail(apiBreeds))
      .finally(() => setLoading(false));
  };

  return (
    <BreedContext.Provider
      value={{
        breeds,
        filteredBreeds,
        searchText,
        breedDetail,
        handleChange,
        handleDetail,
        loading,
      }}
    >
      {children}
    </BreedContext.Provider>
  );
};

export const useBreeds = () => {
  return useContext(BreedContext)?.breeds;
};

export const useFilteredBreeds = () => {
  return useContext(BreedContext)?.filteredBreeds;
};

export const useSearchText = () => {
  return useContext(BreedContext)?.searchText;
};

export const useBreedDetail = () => {
  return useContext(BreedContext)?.breedDetail;
};

export const useHandleChange = () => {
  return useContext(BreedContext)?.handleChange;
};

export const useHandleDetail = () => {
  return useContext(BreedContext)?.handleDetail;
};

export const useLoading = () => {
  return useContext(BreedContext)?.loading;
};

export default BreedContext;
