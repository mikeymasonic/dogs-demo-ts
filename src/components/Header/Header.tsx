import React from 'react';
import { useHandleChange, useSearchText } from '../../hooks/breedProvider';

const Header = () => {
  const searchText = useSearchText();
  const handleChange = useHandleChange();
  const refreshPage = () => {
    window.location.reload();
  };

  return (
    <section>
      <h1 onClick={refreshPage}>Dogs!</h1>
      <input
        type="text"
        placeholder="Search"
        value={searchText}
        onChange={handleChange}
      />
    </section>
  );
};

export default Header;
