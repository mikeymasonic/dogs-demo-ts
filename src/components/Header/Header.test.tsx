import React from 'react';
import { shallow } from 'enzyme';
import * as BreedContext from '../../hooks/breedProvider';
import Header from './Header';

describe('Header component', () => {
  test('it should mock the context and render Header', () => {
    const contextValues = 'husky';

    jest
      .spyOn(BreedContext, 'useSearchText')
      .mockImplementation(() => contextValues);

    jest
      .spyOn(BreedContext, 'useHandleChange')
      .mockImplementation(() => () => {});

    const wrapper = shallow(<Header />);
    expect(wrapper).toMatchSnapshot();
  });
});
