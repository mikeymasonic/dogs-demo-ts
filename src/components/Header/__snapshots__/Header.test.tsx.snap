// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component it should mock the context and render Header 1`] = `ShallowWrapper {}`;
